app.js

```diff
@@ -0,0 +1,130 @@
+/*
+    *** Conventions for this program ***
+
+    *** For the POST API ***
+        DATA FORMAT in JSON is {
+                                    "data": [1,2,3,4,5......]
+                               }
+
+    *** For the PATCH API ***
+        DATA FORMAT in JSON is {
+                                    "num": 5
+                               }
+
+    *** Tested the API calls in POSTMAN
+
+
+*/
+
+const express = require('express');
+const app = express();
+app.use(express.json());
+
+const port = 3000;
+const dataSize = 500;
+
+let numArray = [];
+
+/* 
+    Function to find the index where the new number is to be inserted
+    
+    Parameters: number - number to be inserted
+                low    - the low index for the current search
+                high   - the high index for the current search
+    
+    return value: the index to be inserted at
+*/
+const binarySearch = (number, low, high) => {
+    if (high <= low)
+        return (number > numArray[low])? (low + 1): low;
+    
+    let mid = Math.ceil((low + high) / 2);
+
+    if (number == numArray[mid])
+        return mid + 1;
+
+    if (number > numArray[mid])
+        return binarySearch(number, mid + 1, high)
+    
+    return binarySearch(number, low, mid -1);
+
+}
+
+// GET API 
+const sortNumbers = (req, res) => {

```

Comment:
How would you change your solution if this application is 'read-heavy' such that GET is called frequently but POST is only used rarely? Assume we need to optimize this system so GET can be as fast as possible but still fulfill the requirements given for sorting.

Reply:
In the case of 'read-heavy' GET requests and rare POST requests, I would implement the sorting function in the POST method. We would sort the numbers as soon as we get a valid list of numbers and save the results in a variable. This way we can avoid the sorting overhead of O(nlgn) (assuming worst case). Moreove, if we want to optimize GET requests more, we can cache GET results in the client's computer and only send in new data if the data has been modified since last time.